File: src/runner/code-runner.service.ts
import { Injectable } from '@nestjs/common';
import { CodeRunnerResult, LANGUAGE_MAP } from '../util/constant';
import { NotSupportedLanguageException } from '../exception/domain/not-supported-language.exception';
import { CodeRunRequestDto } from './code-run-request.dto';

@Injectable()
export class CodeRunnerService {
  async runCode(
    codeRunRequestDto: CodeRunRequestDto,
  ): Promise<CodeRunnerResult> {
    const { language, code, input, timeLimitSecond } = codeRunRequestDto;
    const LanguageClass = LANGUAGE_MAP.get(language);
    if (LanguageClass === undefined) {
      throw new NotSupportedLanguageException(codeRunRequestDto.language);
    }

    const instance = new LanguageClass(timeLimitSecond, code, input);
    return instance.runCode();
  }
}

File: src/main/java/com/baekjoon_testcase/baekjoon_testcase/service/CodeRunnerService.java
package com.baekjoon_testcase.baekjoon_testcase.service;

import com.baekjoon_testcase.baekjoon_testcase.dto.TestCodeRequest;
import com.baekjoon_testcase.baekjoon_testcase.dto.TestCodeResponse;

public interface CodeRunnerService {
    TestCodeResponse runCode(TestCodeRequest testCodeRequest);
}
